core/src/main/java/org/apache/iceberg/SnapshotProducer.java

```diff
@@ -257,10 +273,48 @@ public Snapshot apply() {
           .run(index -> manifestFiles[index] = manifestsWithMetadata.get(manifests.get(index)));
 
       writer.addAll(Arrays.asList(manifestFiles));
+
+      if (manifestListKeyMetadata != null) {
+        writer.close();
```

Comment:
This is going to result in a double close for encrypted manifest lists. Instead, I think this should use a try/finally pattern where `close` is called in `finally` instead of using try-with-resources.